File: solutions/triangle/Python/triangle.py
""""Checks if the 3 sides form a valid triangle."""

#class InvalidTriangle(Exception):
#    pass

def valid_triangle(f):
    def check(sides):
        if not all(sides) or not all(side * 2 < sum(sides) for side in sides):
            return False
        return f(sides)
    return check

@valid_triangle    
def equilateral(sides):
    return len({*sides}) == 1

@valid_triangle
def isosceles(sides):
    return len({*sides}) <= 2

@valid_triangle
def scalene(sides):
    return len({*sides}) == 3

File: solutions/triangle/C++/triangle.cpp
#include "triangle.h"
#include <vector>
#include <algorithm>

namespace triangle {
    flavor kind(double a, double b, double c) {
        std::vector<double> sides {a, b, c};
		std::sort(sides.begin(), sides.end());

        if (sides[0] <= 0)
			throw std::domain_error("All sides should have positive length");
		if (sides[0] + sides[1] < sides[2])
			throw std::domain_error("Illegal triange: Violates length inequality");
		if (sides[0] == sides[2])
			return flavor::equilateral;
		if (sides[0] == sides[1] || sides[1] == sides[2])
			return flavor::isosceles;
		return flavor::scalene;        
    }
}
